Don't suppress all 'os.makedirs' exceptions.

elektito opened the pull request:
os.makedirs can raise a number of exceptions, one of which (already
exists) could safely be ignored while we were silencing all possible
exceptions.

As an example of an issue that this could cause, if the ".conan"
directory was located in a directory the user doesn't have write
permission to, conans.util.files.save would raise a FileNotFoundError,
which is misleading.

Fixes #7281.

Changelog: (Feature | Fix | Bugfix): Describe here your pull request
Docs: https://github.com/conan-io/docs/pull/XXXX

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=7284) <br/>All committers have signed the CLA.

elektito committed:
Don't suppress all 'os.makedirs' exceptions.

os.makedirs can raise a number of exceptions, one of which (already
exists) could safely be ignored while we were silencing all possible
exceptions.

As an example of an issue that this could cause, if the ".conan"
directory was located in a directory the user doesn't have write
permission to, conans.util.files.save would raise a FileNotFoundError,
which is misleading.

Fixes #7281.

memsharded commented on conans/util/files.py:
```diff
@@ -161,10 +161,8 @@ def _generic_algorithm_sum(file_path, algorithm_name):
-        os.makedirs(os.path.dirname(path))
-    except Exception:
-        pass
+    if os.path.dirname(path) != '':
```
Try this:

```python
dirname = os.path.dirname(path)
if dirname and not os.path.exists(dirname):  # compat Python 2.7, otherwise exists_ok=True
    try:
       os.makedirs(dirname)
    except Exception as e:
       raise ConanException("Couldn't create folder '%s': %s" % (dirname, str(e))

```

elektito replied:
The `mkdir` function already checks for the existence of the folder. So I don't think we need to do that here.

I'll add the ConanException part right away.

elektito replied:
Updated the PR. I took the liberty of using the format function instead of the old string interpolation style. And like I said, I didn't include the check for existence of the folder, so it will be done in `mkdir`.

What do you think of doing all this in mkdir? Having two copies of the same code in `save` and `save_append` really bugs me, but I don't know how changing `mkdir` can affect the rest of the codebase.

memsharded commented:
This was stuck for a long time, I guess because of being red in CI, sorry about that.

I have tested this behavior for 2.0, and the traceback doesn't appear any more, but a more clear error message saying such folder doesn't have permissions, so I am considering this one superseded by https://github.com/conan-io/conan/pull/12806 and solved in 2.0.

Many thanks!
